### ISSUE TITLE
Library not loaded: @rpath/Charts.framework/Charts Xcode 6.3.1, iOS 8.3

### ISSUE THREAD

**Issue by markleeyw** [danielgindi/ios-charts#57]
Simulator Works fine but when running on an iPhone I get this error.
Xcode 6.3.1 and iOS 8.3 Deployment. 
Any help? 

dyld: Library not loaded: @rpath/Charts.framework/Charts
  Referenced from: /private/var/mobile/Containers/Bundle/Application/F1F8EA37-A2F3-4078-8F65-9291C708B698/stepz.app/stepz
  Reason: image not found
(lldb)

**Comment by danielgindi** [comment]
The framework should be "Embedded". It's in the readme.

**Comment by markleeyw** [comment]
I followed these instructions but still get this error. 
1. Drag the `Charts.xcodeproj` to your project  
2. Go to your target's settings, hit the "+" under the "Embed Frameworks" section, and select the Charts.framework  
3. `@import Charts`  
Please see screenshot. 
![error](https://cloud.githubusercontent.com/assets/12169084/7389978/dd1754f0-eea5-11e4-9686-40134905d73a.png)

**Comment by danielgindi** [comment]
"Embedded Binaries"!

**Comment by markleeyw** [comment]
Yup its working now ... sorry silly mistake !! ha ha !

**Comment by mseijas** [comment]
I'm having a similar problem... Every time I add the framework (either as embedded binary and/or linked), the framework shows in red as if it weren't able to find it.
I can @import Charts, and declare Classes, but when I run the simulator they just don't work at all... Or through exceptions when I try to modify some of their properties..

Any ideas? I'm really at a loss here.

Thanks!

**Comment by liuxuan30** [comment]
@mseijas you might need to provide more error logs. I am not clear what's your problem. Can you compile it? Can you create the chart objects? what is `framework shows in red as if it weren't able to find it.`?

Check build settings about swift, and follow the README exactly.

If it's Objective-C based project, set `embedded content contains swift code` to YES, `define modules` to YES, and the `module name` is correct, also make sure `Objective-C generated interface header file` name is correct.

**Comment by mseijas** [comment]
I'm working on an Objectve-C based project with iOS 8. 

- I've imported the Charts.xcodeproj into my project
- Compiled it (successfully)
- Created a bridging header for my project
- Set "Embedded Content Contains Swift Code" to YES

But every reference to "Charts.xcodeproj" appears in red
![Screenshot](http://i.imgur.com/biPUsOF.png "Screenshot")

I can do `@import Charts` into a View Controller --> Complies and Runs

I can declare a UIView as a `LineChartView` --> Compiles and Runs, but no "No Data" message appears, making me believe that it's not really creating a LineChartViewObject

And then, if I try to do `lineChartView.noDataText = @"No Data Available"` --> Compiles, but throws a runtime exception:

```
2015-06-25 09:20:39.257 CP3[7433:2119612] -[UIView setNoDataText:]: unrecognized selector sent to instance 0x7fb118552300
2015-06-25 09:20:39.259 CP3[7433:2119612] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[UIView setNoDataText:]: unrecognized selector sent to instance 0x7fb118552300'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000103cd7c65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000103970bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x0000000103cdf0ad -[NSObject(NSObject) doesNotRecognizeSelector:] + 205
	3   CoreFoundation                      0x0000000103c3513c ___forwarding___ + 988
	4   CoreFoundation                      0x0000000103c34cd8 _CF_forwarding_prep_0 + 120
	5   CP3                                 0x000000010305c9f1 -[ViewController viewDidLoad] + 129
	6   UIKit                               0x0000000104201210 -[UIViewController loadViewIfRequired] + 738
	7   UIKit                               0x000000010420140e -[UIViewController view] + 27
	8   UIKit                               0x000000010411c2c9 -[UIWindow addRootViewControllerViewIfPossible] + 58
	9   UIKit                               0x000000010411c68f -[UIWindow _setHidden:forced:] + 247
	10  UIKit                               0x0000000104128e21 -[UIWindow makeKeyAndVisible] + 42
	11  UIKit                               0x00000001040cc457 -[UIApplication _callInitializationDelegatesForMainScene:transitionContext:] + 2732
	12  UIKit                               0x00000001040cf1de -[UIApplication _runWithMainScene:transitionContext:completion:] + 1349
	13  UIKit                               0x00000001040ce0d5 -[UIApplication workspaceDidEndTransaction:] + 179
	14  FrontBoardServices                  0x0000000105dba5e5 __31-[FBSSerialQueue performAsync:]_block_invoke_2 + 21
	15  CoreFoundation                      0x0000000103c0b41c __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12
	16  CoreFoundation                      0x0000000103c01165 __CFRunLoopDoBlocks + 341
	17  CoreFoundation                      0x0000000103c00f25 __CFRunLoopRun + 2389
	18  CoreFoundation                      0x0000000103c00366 CFRunLoopRunSpecific + 470
	19  UIKit                               0x00000001040cdb42 -[UIApplication _run] + 413
	20  UIKit                               0x00000001040d0900 UIApplicationMain + 1282
	21  CP3                                 0x000000010305cddf main + 111
	22  libdyld.dylib                       0x000000010a33e145 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```


Any ideas? Thank you so much for your help!!

**Comment by danielgindi** [comment]
But Xcode told you already! I don't understand why you are asking...
It says "`reason: '-[UIView setNoDataText:]: unrecognized selector sent to instance 0x7fb118552300`"...

In other words, you are sending a `setNoDataText` message to a `UIView`. Now why would you do that? `UIView` has no idea how to respond to that! You have offended it!

If you really want to send such messages, you better try your luck with a `ChartView` like, for example, a `LineChartView`. He is more mature and can definitely handle it!

**Comment by danielgindi** [comment]
Or maybe She?

**Comment by mseijas** [comment]
Hahaha! No, but that's the thing... I actually haven't defined it as a UIView but as a LineChartView... except that I don't think Xcode is picking up on that... 

Look!

`@property (weak, nonatomic) IBOutlet LineChartView *chartView;`

That's the object i'm sending the `setNoDataText` message to

**Comment by danielgindi** [comment]
No friend, that's not the object. That's a definition of a pointer. The
question is what object have you set it to point to?
>
>
Also, on a side note: It's a weak pointer. Which means that if you
temporarily remove the object from your view, the pointer will be set to
null and you won't be able to return it to the screen.

**Comment by mseijas** [comment]
OH.MY.GOD!! You're so right!! I've been beating my head against the wall for 3 days to figure this out, can't believe I miss that!

THANK YOU!! Thank you SO SO much!!

**Comment by danielgindi** [comment]
You're welcome and have a nice day ;-)

**Comment by abraxascorner** [comment]
I have a similar problem. My app works when I run in simulator, but when I deploy for in house distribution, it crashes. 
```
Dyld Error Message:
Dyld Message: Library not loaded: @rpath/Charts.framework/Charts
  Referenced from: /var/mobile/Containers/Bundle/Application/29E15804-2B5B-4C58-A9CF-CC0EEF80D758/Moj Telemach.app/Moj Telemach
  Reason: no suitable image found.  Did find:
	/private/var/mobile/Containers/Bundle/Application/29E15804-2B5B-4C58-A9CF-CC0EEF80D758/Moj Telemach.app/Frameworks/Charts.framework/Charts: mmap() errno=1 validating first page of '/private/var/mobile/Containers/Bundle/Application/29E15804-2B5B-4C58-A9CF-CC0EEF80D758/Moj Telemach.app/Frameworks/Charts.framework/Charts'
  Dyld Version: 369.12

Binary Images:
0x1000f4000 - 0x100263fff Moj Telemach arm64  <3ed6ff1c15f531ffaaddb1111ac253d1> /var/mobile/Containers/Bundle/Application/29E15804-2B5B-4C58-A9CF-CC0EEF80D758/Moj Telemach.app/Moj Telemach
0x12006c000 - 0x120097fff dyld arm64  <1b945967d665331dbf2600249e46660a> /usr/lib/dyld
```

**Comment by danielgindi** [comment]
Seems like you have a linking problem (Framework was not embedded
correctly).

Please search for many posts regarding this... Or follow the README
carefully.

**Comment by abraxascorner** [comment]
@danielgindi  thanks for help (I already followed readme file). 
But here is what solved my problem (if someone comes across this thread):
if you have cert that is created before swift (b.s. :)) and you want to do in house distribution using swift code in app, you need to revoke cert. Revoke cert and check if it containts Organizational Unit in subject part (keychain assistant -> right click info Get info on cert). There is a lot more about this problem here:
https://github.com/danielgindi/ios-charts/issues/154
Best regards, Nemanja

**Comment by danielgindi** [comment]
You do not have to revoke your certificate, that is a mistake.
You only need to re-download your provisioning profile, and of course
delete all of the old profiles from your mac to make sure that you are
using the correct one.

‏בתאריך יום חמישי, 1 באוקטובר 2015, abraxascorner <notifications@github.com>
כתב:

**Comment by abraxascorner** [comment]
@danielgindi I created provision profile today, new one, but it didn't work. Then I revoked distribution cert and I was able to run app without crashing on start.

**Comment by iOSArchitect** [comment]
Hi,

I am facing libswiftCore.dylib: bitcode_strip failed with exit code 64 issue, I don't know how to get rid of this problem, Could you please guide me?

Thanks,

<img width="726" alt="screen shot 2015-12-22 at 12 21 24 am" src="https://cloud.githubusercontent.com/assets/14069953/11941625/0bc10980-a842-11e5-83ed-188cd39492ca.png">

**Comment by danielgindi** [comment]
It seems to fail when optimizing the Swift library for your binary. "Copying libSwiftCore..."

So its some Xcode bug. Maybe the libraries are not downloaded correctly from the AppStore, or maybe a permissions problem on Xcode's temporary folders...

**Comment by liuxuan30** [comment]
Do not post same question in old issues. If you had problem that is different, file a new one.
To me, it's saying `--toolchain` is unrecoginized option. It smells like a proejct setting issue. check it first.

It should be easy to triage: /usr/bin/xcrun is the program, and --toolchain cannot be recognized by /usr/bin/xcrun. You can start from here. Use other machines /usr/bin/xcrun, manually run /usr/bin/xcrun, check version, anything.

**Comment by lzantal** [comment]
Hi all,

Had this same issue. I fixed it easily by removing it from embedded binaries and just drag the folder in with all the swift classes. First had to build it on the device and took awhile to compile everything. 
After that I was able to easily do an in-house build without any other changes to the project and build settings.

**Comment by TheCodeor** [comment]
Thank you, because of his carelessness did not carefully read README, attention next time!

**Comment by RagulLakshmanan** [comment]
dyld: Library not loaded: @rpath/Charts.framework/Charts
  Referenced from: /var/containers/Bundle/Application/C282964C-2B51-4810-9544-201E56ECD0A5/AwaasV1.app/AwaasV1
  Reason: no suitable image found.  Did find:
	/private/var/containers/Bundle/Application/C282964C-2B51-4810-9544-201E56ECD0A5/AwaasV1.app/Frameworks/Charts.framework/Charts: mmap() errno=1 validating first page of '/private/var/containers/Bundle/Application/C282964C-2B51-4810-9544-201E56ECD0A5/AwaasV1.app/Frameworks/Charts.framework/Charts'


can any one pls help me with this.. Got stuck for more than a day. I have added embedded libraries too. Please help meeeee

**Comment by liuxuan30** [comment]
@RagulLakshmanan the solution is all here, and google and stack overflow. I'd say take rest and start over again.

**Comment by lzantal** [comment]
@RagulLakshmanan did you try my fix I mentioned above? I just used it last week for another project and everything works as expected.

**Comment by RagulLakshmanan** [comment]
@lzantal  @liuxuan30.. Thank you all.. The issue was the certificates. I have cleared all the certificates and provisioning profile and kept only the required one. Having profiles and cert in one file will cause this problem. The code is working and took build in device too. Happy coding.

**Comment by geoherna** [comment]
Incase this helps anyone, none of the solutions I kept finding on the web were working for me. Pulled my hair our for 2 days, and tried everything. I revoked in-house cert, new provisioning profile, added files to embedded, etc.

Could not for the life of me figure out what was wrong until I noticed that in Keychain access my Apple WWDR and iOS Distributions certs were being set to **"Always Trust"** instead of "Use Systems Defaults". Switched my certs back to **"Use Systems Defaults"** and everything went back to working as it should. I have no idea why and how this works but it did.
